02-microarray/differential-expression_microarray_02_several-groups.Rmd

```diff
@@ -0,0 +1,456 @@
+```
+
+Now let's create the results table based on the contrasts fitted model. 
+
+This step will provide the Benjamini-Hochberg multiple testing correction. 
+The `topTable()` function default is to use Benjamini-Hochberg but this can be changed to a different method using the `adjust.method` argument (see the `?topTable` help page for more about the options).
+
+```{r}
+# Apply multiple testing correction and obtain stats
+stats_df <- topTable(contrasts_fit, number = nrow(df)) %>%
+  tibble::rownames_to_column("Gene")
+```
+
+Let's take a peek at what our results table looks like. 
+
+```{r}
+head(stats_df)
+```
+
+Each group's fold change in comparison to the other groups average are reported. 
+
+By default, results are ordered by largest `F` value to the smallest, which means your most differentially expressed genes should be toward the top. 
+
+See the help page by using `?topTable` for more information and options for this table.
+
+## Check results by plotting one gene
+
+To test if these results make sense, we can make a plot of one of top genes. 
+Let's try extracting the data for `ENSG00000128683` and set up its own data frame for plotting purposes. 
+Based on the results in `stats_df`, we should expect this gene to be much higher in the `WNT` samples. 
+
+First we will need to set up the data for this gene and the subgroup labels into a data frame for plotting. 
+
+```{r}
+top_gene_df <- df %>%
+  # Extract this gene from `df`
+  dplyr::filter(rownames(.) == "ENSG00000128683") %>%
+  # Transpose so the gene is a column
+  t() %>%
+  # Transpose made this a matrix, let's make it back into a data.frame like before
+  data.frame() %>%
+  # Store the sample ids as their own column instead of being row names
+  tibble::rownames_to_column("refinebio_accession_code") %>%
+  # Join on the selected columns from metadata
+  dplyr::inner_join(dplyr::select(
+    metadata,
+    refinebio_accession_code,
+    subgroup
+  ))
+```
+
+Let's take a sneak peek at what our `top_gene_df` looks like. 
+
+```{r}
+top_gene_df
+```
+
+Now let's plot the data for `ENSG00000128683` using our `top_gene_df`.
+We should expect this gene to be expressed at much higher levels in the `WNT` group samples. 
+
+```{r}
+ggplot(top_gene_df, aes(x = subgroup, y = ENSG00000128683, color = subgroup)) +
+  geom_jitter(width = 0.2) + # We'll make this a jitter plot
+  theme_classic() # This makes some aesthetic changes
+```
+
+Yes! These results make sense.
+The WNT samples have much higher expression of ENSG00000128683 than the other samples. 
+
+## Write results to file
+
+The results in `stats_df` will be saved to our `results/` directory. 
+
+```{r}
+readr::write_tsv(stats_df, file.path(
+  results_dir,
+  "GSE37418_limma_results.tsv" # Replace with a relevant output name
+))
+```
+
+## Make a volcano plot
+
+We'll use the `EnhancedVolcano` package's main function to plot our data [@Zhu2018].
+
+We will have to choose one of the groups fold changes to plot. 
```

Comment:
I can't exactly think of what the better plot would be; It might be something where you plotted lines between the 4 groups to try to tease out what patterns are most common. Something like this?
```
plot_df <- stats_df %>% 
  dplyr::filter(adj.P.Val < 1e-20) %>%
  tidyr::pivot_longer(cols = ends_with("vsOther"), 
                      names_to = "contrast", 
                      values_to = "logFoldChange")

ggplot(plot_df, aes(x = contrast, y = logFoldChange, group = Gene)) + 
  geom_line(alpha = 0.1)
```

I don't love that, and might want some other kinds of filtering, but it is *something*

Reply:
I definitely agree with your point about the p value. I don't love this line plot either. I'll see if I can find/come up with a different visual that makes sense.

Reply:
This is still the ANOVA p value, so I'm not sure the problem is improved, except that it is clearer that it is for the ANOVA and all groups are represented. 

<img width="706" alt="Screen Shot 2020-09-24 at 4 49 18 PM" src="https://user-images.githubusercontent.com/23458084/94198862-f30c5a80-fe85-11ea-9d43-45c174850089.png">

Reply:
Well, I'm not sure it is technically an ANOVA, so don't say that. But it is ANOVA like in the sense that ANOVA is just a special linear model (Everything is a linear model). I think there is a way to get the individual contrast p values out, but it isn't with `topTable()` I can't remember right now.

Reply:
Here's a same idea as what you original posted, but more zazz. 
<img width="713" alt="Screen Shot 2020-09-24 at 5 20 58 PM" src="https://user-images.githubusercontent.com/23458084/94201674-54cec380-fe8a-11ea-88ea-79fd304b6f96.png">

Reply:
Extracting the contrast p values was not to bad, there's some more data wrangling steps now because of that but I think that's a good thing to show more examples of.

Reply:
Cool. If you want to do the more complicated analysis, expect it to get more complicated...